ow_power_system/src/PowerSystemPack.cpp

```diff
@@ -0,0 +1,375 @@
+#include <std_msgs/Float64.h>
+#include <std_msgs/Int16.h>
+#include <ow_lander/lander_joints.h>
+
+#include <ros/ros.h>
+#include <sensor_msgs/JointState.h>
+#include <PrognoserFactory.h>
+
+#include "Messages/MessageBus.h"
+#include "Messages/ProgEventMessage.h"
+#include "Messages/ScalarMessage.h"
+#include "ModelBasedAsyncPrognoserBuilder.h"
+
+#include "PowerSystemPack.h"
+#include "PowerSystemNode.h"
+
+using namespace PCOE;
+
+const auto START_TIME       = MessageClock::now();
+//const int NUM_NODES         = 8; // NOTE: Should equal m_num_nodes in PowerSystemPack.h
+
+// The index use to access temperature information.
+// This might change to median SOC or RUL index or fixed percentile.
+//
+static constexpr int TEMPERATURE_INDEX = 1;
+
+double EoD_events[NUM_NODES][3];
+
+// The PredictionHandler class subscribes to the battery EoD event message and
+// prints each event as it is received.
+class PredictionHandler : public IMessageProcessor 
+{
+public:
+  /**
+   * Constructs a new prediction handler that subscribes to battery EoD
+   * predictions for the specified source and on the specified message
+   * bus.
+   **/
+  PredictionHandler(MessageBus& bus, const std::string& src) : bus(bus)
+  {
+    bus.subscribe(this, src, MessageId::BatteryEod);
+    identifier = src;
+    node_number = src[5] - '0'; // converts node identifier value to int
+  }
+
+  /**
+   * Unsubscribes the prediction handler from the message bus.
+   **/
+  ~PredictionHandler()
+  {
+    bus.unsubscribe(this);
+  }
+
+  /**
+   * The message bus will call this function each time the predictor publishes
+   * a new battery EoD prediction.
+   **/
+  void processMessage(const std::shared_ptr<Message>& message) override
+  {
+    using namespace std::chrono;
+    // The prediction handler only ever subscribes to the BatteryEoD message
+    // id, which should always be a ProgEventMessage, so this should always
+    // succeed.
+    auto prediction_msg = dynamic_cast<ProgEventMessage*>(message.get());
+    if (prediction_msg == nullptr)
+    {
+      std::cerr << "Failed to cast prediction message to expected type" << std::endl;
+      std::exit(1);
+    }
+
+    // Get the event for battery EoD
+    auto eod_event = prediction_msg->getValue();
+
+    // The time of event is a `UData` structure, which represents a data
+    // point while maintaining uncertainty. For the MonteCarlo predictor
+    // used by this example, the uncertainty is captured by storing the
+    // result of each particle used in the prediction.
+    UData eod_time = eod_event.getTOE();
+    if (eod_time.uncertainty() != UType::Samples)
+    {
+      std::cerr << "Unexpected uncertainty type for EoD prediction" << std::endl;
+      return std::exit(1);
+    }
+
+    // valid prediction
+    // Determine the median RUL.
+    auto samplesRUL = eod_time.getVec();
+    sort(samplesRUL.begin(), samplesRUL.end());
+    double eod_median = samplesRUL.at(samplesRUL.size() / 2);
```

Comment:
This code regarding the median RUL/SoC/battery temp calculations was taken from the previous writer, so I didn't modify it at first in case it caused issues with the output that I wouldn't be able to detect. It should be simple enough to update though, so I'll take a look.

Reply:
I wrote a simple function to find the median using `nth_element`, but until I get a confirmation on whether or not the original behavior was intended, I'll keep using the old code. I'll continue going through this review tomororw.